fe/fe-core/src/main/java/com/starrocks/scheduler/MvTaskRunProcessor.java

```diff
@@ -159,36 +166,84 @@ public void processTaskRun(TaskRunContext context) {
+                        partitionTable, needRefreshPartitionNames);
             }
         }
     }
+    public Map<Long, OlapTable> collectBaseTables(MaterializedView materializedView, Database database) {
+        Map<Long, OlapTable> olapTables = Maps.newHashMap();
+        Set<Long> baseTableIds = materializedView.getBaseTableIds();
+        database.readLock();
+        try {
+            for (Long baseTableId : baseTableIds) {
+                OlapTable olapTable = (OlapTable) database.getTable(baseTableId);
+                if (olapTable == null) {
+                    throw new SemanticException("Materialized view base table: " + baseTableId + " not exist.");
+                }
+                OlapTable copied = new OlapTable();
+                if (!DeepCopy.copy(olapTable, copied, OlapTable.class)) {
+                    throw new SemanticException("Failed to copy olap table: " + olapTable.getName());
+                }
+                olapTables.put(olapTable.getId(), copied);
+            }
+        } finally {
+            database.readUnlock();
+        }
+        return olapTables;
+    }
 
     private void processPartitionWithPartitionTable(Database database, MaterializedView materializedView,
                                                     OlapTable olapTable, Expr partitionExpr,
                                                     Column partitionColumn, Map<String, String> partitionProperties,
                                                     DistributionDesc distributionDesc,
                                                     Set<String> needRefreshPartitionNames) {
-        // used to get delete table partitions
+        // 0. prepare get need add partition
         long baseTableId = olapTable.getId();
         Set<String> partitionNames = Sets.newHashSet();
+        Set<Partition> needAddPartitions = Sets.newHashSet();
         ExpressionRangePartitionInfo expressionRangePartitionInfo =
                 (ExpressionRangePartitionInfo) materializedView.getPartitionInfo();
         RangePartitionInfo baseRangePartitionInfo = (RangePartitionInfo) olapTable.getPartitionInfo();
         Collection<Partition> basePartitions = olapTable.getPartitions();
         for (Partition basePartition : basePartitions) {
-            long basePartitionId = basePartition.getId();
-            String basePartitionName = basePartition.getName();
             // record full partitions
-            partitionNames.add(basePartitionName);
+            partitionNames.add(basePartition.getName());
             // if exists, not check
             if (!materializedView.needAddBasePartition(baseTableId, basePartition)) {
                 continue;
             }
-            materializedView.addBasePartition(baseTableId, basePartition);
+            needAddPartitions.add(basePartition);
+        }
+        // 1. remove partition refs & get need refresh mv partition ids
```

Comment:
is there any difference if remove partition first, and then add new paprtitions?

Reply:
If the partition is added first, it may conflict with the existing partition range and throw an exception